src/KOKKOS/pair_snap_kokkos_impl.h

```diff
@@ -632,201 +698,145 @@ void PairSNAPKokkos<DeviceType>::operator() (TagPairSNAPComputeNeigh,const typen
-        const auto jdxb = jcoeff % idxb_max;
-        const auto jdx_chem = jcoeff / idxb_max;
-        double bvecj = my_sna.blist(jdxb, jdx_chem, ii);
-        d_beta_pack(iatom_mod,icoeff,iatom_div) += d_coeffi[k]*bvecj;
-        d_beta_pack(iatom_mod,jcoeff,iatom_div) += d_coeffi[k]*bveci;
-        k++;
-      }
-    }
-  }
+  my_sna.compute_cayley_klein(iatom_mod,jnbor,iatom_div);
 }
 
-template<class DeviceType>
+template<class DeviceType, typename real, int vector_length>
 KOKKOS_INLINE_FUNCTION
-void PairSNAPKokkos<DeviceType>::operator() (TagPairSNAPPreUi,const typename Kokkos::TeamPolicy<DeviceType,TagPairSNAPPreUi>::member_type& team) const {
-  SNAKokkos<DeviceType> my_sna = snaKK;
+void PairSNAPKokkos<DeviceType, real, vector_length>::operator() (TagPairSNAPPreUi, const int iatom_mod, const int j, const int iatom_div) const {
+  SNAKokkos<DeviceType, real, vector_length> my_sna = snaKK;
 
-  // Extract the atom number
-  const int ii = team.team_rank() + team.team_size() * (team.league_rank() % ((chunk_size+team.team_size()-1)/team.team_size()));
+  const int ii = iatom_mod + iatom_div * vector_length;
   if (ii >= chunk_size) return;
+
   int itype = type(ii);
   int ielem = d_map[itype];
 
-  my_sna.pre_ui(team,ii,ielem);
+  my_sna.pre_ui(iatom_mod, j, ielem, iatom_div);
 }
 
-template<class DeviceType>
+template<class DeviceType, typename real, int vector_length>
 KOKKOS_INLINE_FUNCTION
-void PairSNAPKokkos<DeviceType>::operator() (TagPairSNAPComputeUi,const typename Kokkos::TeamPolicy<DeviceType,TagPairSNAPComputeUi>::member_type& team) const {
-  SNAKokkos<DeviceType> my_sna = snaKK;
+void PairSNAPKokkos<DeviceType, real, vector_length>::operator() (TagPairSNAPComputeUi,const typename Kokkos::TeamPolicy<DeviceType,TagPairSNAPComputeUi>::member_type& team) const {
+  SNAKokkos<DeviceType, real, vector_length> my_sna = snaKK;
 
-  // Extract the atom number
-  int ii = team.team_rank() + team.team_size() * (team.league_rank() % ((chunk_size+team.team_size()-1)/team.team_size()));
-  if (ii >= chunk_size) return;
+  constexpr int team_size = sizeof(real) == 4 ? 8 : 4;
+  if (team.team_size() != team_size) return; // error
 
-  // Extract the neighbor number
-  const int jj = team.league_rank() / ((chunk_size+team.team_size()-1)/team.team_size());
-  const int ninside = d_ninside(ii);
-  if (jj >= ninside) return;
+  // extract flattened atom_div / neighbor number / bend location
+  int flattened_idx = team.team_rank() + team.league_rank() * team_size;
+
+  // extract neighbor index, iatom_div
+  const int iatom_div = flattened_idx / (max_neighs * (twojmax + 1)); // plug in int_fastdiv
+  const int jj_jbend = flattened_idx - iatom_div * (max_neighs * (twojmax + 1));
+  const int jbend = jj_jbend / max_neighs;
+  const int jj = jj_jbend - jbend * max_neighs;
+
+  // check base atom numbers, hopefully this doesn't lead to blocking conditions?
+  // yep, we can't have a return in here
```

Comment:
Please clean up comments.

Reply:
Can do, sorry, I thought I'd cleaned these up.

Reply:
Done, did an additional sweep for sloppy comments too.